### PR TITLE
VGG tutorial

### PR DESCRIPTION
**Pull request by GregorySenay**
This tutorial explains how to create your own image classifier.

### CODE REVIEW

File: content/vgg-classifier-tutorial.md [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+      batch_size=64,
+      validation_split=0.1,
+      nb_epoch=300,
+      show_accuracy=True,
+      shuffle=True)
+
+mymodel.save_weights('my_model_weights.h5')
+```
+
+If you respect this tutorial, you can quickly achieve an accuracy higher than 85% on caltech101 in 30 epochs, but don't hesitate to play with the different parameters of the model, like the batch_size, learning_rate... to improve the final validation accuracy.
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: "If you _follow_ this tutorial"

The parameters I would suggest tuning would be the size and number of FC layers after the frozen convolution layers.

I think we need to introduce the possibility of fine-tuning the entire network (after de-freezing the convolution layers) with a very low learning rate, to improve final accuracy. The effect of this is huge.

File: content/vgg-classifier-tutorial.md [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+* shuffle=True, to shuffle the data at each epoch
+
+And at the end save the new model weights.
+
+```python
+mymodel.fit(X_data, Y_data,
+      batch_size=64,
+      validation_split=0.1,
+      nb_epoch=300,
+      show_accuracy=True,
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: The `show_accuracy` argument is deprecated in Keras 1.0.

File: content/vgg-classifier-tutorial.md [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+  else:
+    layer.trainable = True # Unlocked for the Lasts
+```
+### Compile the model
+Now, the model can be compile with the categorical cross-entropy loss function  (typically of classification model).
+Note that the learning rate need to be low, else the system is not able to adapted the model for your dataset.
+If it doesn't work for your own dataset, try to decrease first your the learning rate ``lr`` or next increase it (i.e. 0.05) or change the optimizer by `AdaGrad`, `AdaDelta`...
+```python
+sgd = SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
+mymodel.compile(optimizer=sgd, loss='categorical_crossentropy')
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: In Keras 1.0, we need to pass an argument `metrics=["accuracy"]` to monitor accuracy.

File: content/vgg-classifier-tutorial.md [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    layer.trainable = False # Locked for the Firsts
+  else:
+    layer.trainable = True # Unlocked for the Lasts
+```
+### Compile the model
+Now, the model can be compile with the categorical cross-entropy loss function  (typically of classification model).
+Note that the learning rate need to be low, else the system is not able to adapted the model for your dataset.
+If it doesn't work for your own dataset, try to decrease first your the learning rate ``lr`` or next increase it (i.e. 0.05) or change the optimizer by `AdaGrad`, `AdaDelta`...
+```python
+sgd = SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md:279]: I suggest doing training with `rmsprop` or `adadelta`, which do not require parameter tuning, whereas SGD is difficult to tune.

File: content/vgg-classifier-tutorial.md [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+```
+and reattached it to the before last layer:
+
+```python
+mymodel.layers[-1].set_previous(pretrained_model.layers[-2])
+```
+Nothing more !
+
+### Lock some layers for training
+Because ImageNet is a giant dataset compare to Caltech, the VGG16 pretrained model already know how to recognize the image details, but the model must be adapted/tuned to this dataset. A simple way to do this is to lock  all the convolution layers during the training, meaning the weights of the convolutional layers are never be update. Only the fully connected layers (``Dense``) are updated. But you can easily make some tests by changing the value of: ``unlock_last``. Increasing ``unlock_last`` means: more layers with by updated, decreasing ``unlock_last`` means: less layers are updated.
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: "Because ImageNet is a giant dataset _compared_ to Caltech"

"the VGG16 pretrained model already _knows_"

This part needs more work, to explain how the VGG network already has learned appropriate representations, but that our small dataset would cause quick overfitting the these representations. This is the reason why we need to freeze (the term is freeze, not lock) the convolution layers.

File: content/vgg-classifier-tutorial.md [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+
+```python
+mymodel.layers[-1] = Dense(101, activation='softmax'))
+```
+and reattached it to the before last layer:
+
+```python
+mymodel.layers[-1].set_previous(pretrained_model.layers[-2])
+```
+Nothing more !
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: Not space before `!` (english conventions).

File: content/vgg-classifier-tutorial.md [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+### Change the last layer
+For changing the number of classes of your model, you need to redefine the layer:
+
+```python
+mymodel.layers[-1] = Dense(101, activation='softmax'))
+```
+and reattached it to the before last layer:
+
+```python
+mymodel.layers[-1].set_previous(pretrained_model.layers[-2])
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: This will not work with the new Keras. I am 99% sure that the following would work, although it needs to be tested"

``` python
del model.layers[-1]
model.add(Dense(101, activation='softmax'))
```

I would suggest deleting all FC layers are replacing them with new, smaller FC layers (adapted to a small dataset).

File: content/vgg-classifier-tutorial.md [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+	im = im.transpose((2,0,1))
+	#im = np.expand_dims(im, axis=0)
+	return im
+```
+
+Once you're done you can save the file in the numpy format, to avoid to recreate the dataset every time:
+
+```python
+...
+np.save("cifar100_X",X_data)
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: I thought this was caltech101?

File: content/vgg-classifier-tutorial.md [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+		* image_0001.jpg
+ 		* image_0002.jpg
+		* ...
+ 	* accordion
+	* ...
+``
+
+Advice: if you create your own classes, try to have a minimum of 40 images with a lot of variabilities (i.e. if you have a cat category try to have different cat positions and different cat type, color...).
+
+Unfortunately, the model you are using below requires a same dimension for all of the images, so we need to load and harmonize the image dataset.
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md:57]: "the model we will be using here requires all input images to have a specific size, 224x224".

We should actually remove the FC layers from the original network to make it possible to train on any image size (as long as all new images are the same size). It is fine to do that because in the example below you are deleting the last logreg layer anyway.

File: content/vgg-classifier-tutorial.md [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+		* ...
+ 	* Motorbikes
+		* image_0001.jpg
+ 		* image_0002.jpg
+		* ...
+ 	* accordion
+	* ...
+``
+
+Advice: if you create your own classes, try to have a minimum of 40 images with a lot of variabilities (i.e. if you have a cat category try to have different cat positions and different cat type, color...).
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: `variability`. Why 40? Why not 50? A bit of justification would be welcome.

File: content/vgg-classifier-tutorial.md [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+    * Class3Name
+      * Image1Class1
+      * Image2Class1
+      * ...
+    * Class4Name
+    * ...
+
+```
+
+In the case of Caltech101, (available here: http://www.vision.caltech.edu/Image_Datasets/Caltech101/101_ObjectCategories.tar.gz), once you extract the archive with ``tar xvf 101_ObjectCategories.tar.gz)`` you have this directory architecture:
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: Download link should be introduced in a separate line: "The Caltech101 data is available for download [from this link]." Please use Markdown link syntax to specify hyperlinks.

File: content/vgg-classifier-tutorial.md [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+
+Prepare your dataset
+-----
+Here, I will explain how to simply prepare your image dataset following an existing simple one: Caltech101.
+Caltech101 is a dataset of 101 categories (crocodile, camera, plane, soccer ball...) with 40 to 800 images per category, the size of image is bout 200x300.
+To prepare your data, you have to follow this directory architecture:
+```
+  * Main Directory # Name of you dataset
+    * Class1Name
+      * Image1Class1
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: It would be nice to specify that `Class1Name` is a directory (e.g. by writing it as `Class1Name/`) and specify that `Image1Class1` are picture files (e.g. by writing it as `Image1Class1.jpg`)

File: content/vgg-classifier-tutorial.md [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+Category: Tutorial
+Author: Gregory Senay
+
+
+This tutorial explains how to create image classifier, from an ImageNet pre-trained model, your own model on your own dataset.
+
+Prepare your dataset
+-----
+Here, I will explain how to simply prepare your image dataset following an existing simple one: Caltech101.
+Caltech101 is a dataset of 101 categories (crocodile, camera, plane, soccer ball...) with 40 to 800 images per category, the size of image is bout 200x300.
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: "with 40 to 800 images per category. The size of each image is 200x300." Is it the exact size?

File: content/vgg-classifier-tutorial.md [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+Date: 2016-03-30
+Category: Tutorial
+Author: Gregory Senay
+
+
+This tutorial explains how to create image classifier, from an ImageNet pre-trained model, your own model on your own dataset.
+
+Prepare your dataset
+-----
+Here, I will explain how to simply prepare your image dataset following an existing simple one: Caltech101.
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: "simply" could be removed.

File: content/vgg-classifier-tutorial.md [hunk 15]
```diff
@@ -0,0 +1,320 @@
+Title: Keras, Make your own image classifier
+Date: 2016-03-30
+Category: Tutorial
+Author: Gregory Senay
+
+
+This tutorial explains how to create image classifier, from an ImageNet pre-trained model, your own model on your own dataset.
+
+Prepare your dataset
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: Markdown syntax: titles should be prefixed with `##`

File: content/vgg-classifier-tutorial.md [hunk 16]
```diff
@@ -0,0 +1,320 @@
+Title: Keras, Make your own image classifier
+Date: 2016-03-30
+Category: Tutorial
+Author: Gregory Senay
+
+
+This tutorial explains how to create image classifier, from an ImageNet pre-trained model, your own model on your own dataset.
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: "how to create _an_ image classifier, using a model pre-trained on ImageNet and fine-tuned over your own dataset"

File: content/vgg-classifier-tutorial.md [hunk 17]
```diff
@@ -0,0 +1,320 @@
+Title: Keras, Make your own image classifier
```
Comment (fchollet) [inline content/vgg-classifier-tutorial.md]: I'd prefer "Make your own image classifier with Keras"

**Comment by fchollet** [comment]
Cool, good work! I left comments of the parts that can be improved.

I am noticing that the entire process can be automated. As such, it would be great to have an example script that be run without modifications, from the command line, to create, train and save the model. Just pass it the path to your main data directory, the path to the original VGG16 model, and it does all the work. Such a script should be added to the Keras examples directory.
